zetaclient/chains/solana/observer/outbound.go

```diff
@@ -2,22 +2,344 @@ package observer
+		return err
+	}
+
+	// create outbound ticker based on chain params
+	chainID := ob.Chain().ChainId
+	ticker, err := clienttypes.NewDynamicTicker(
+		fmt.Sprintf("Solana_WatchOutbound_%d", chainID),
+		ob.GetChainParams().OutboundTicker,
+	)
+	if err != nil {
+		ob.Logger().Outbound.Error().Err(err).Msg("error creating ticker")
+		return err
+	}
+
+	ob.Logger().Outbound.Info().Msgf("WatchOutbound started for chain %d", chainID)
+	sampledLogger := ob.Logger().Outbound.Sample(&zerolog.BasicSampler{N: 10})
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C():
+			if !app.IsOutboundObservationEnabled(ob.GetChainParams()) {
+				sampledLogger.Info().Msgf("WatchOutbound: outbound observation is disabled for chain %d", chainID)
+				continue
+			}
+			trackers, err := ob.ZetacoreClient().GetAllOutboundTrackerByChain(ctx, chainID, interfaces.Ascending)
+			if err != nil {
+				continue
+			}
+
+			for _, tracker := range trackers {
+				// go to next tracker if this one already has a finalized tx
+				nonce := tracker.Nonce
+				if ob.IsTxFinalized(tracker.Nonce) {
+					continue
+				}
+
+				// get original cctx parameters
+				cctx, err := ob.ZetacoreClient().GetCctxByNonce(ctx, chainID, tracker.Nonce)
+				if err != nil {
+					// take a rest if zetacore RPC breaks
+					ob.Logger().Outbound.Error().
+						Err(err).
+						Msgf("WatchOutbound: can't find cctx for chain %d nonce %d", chainID, tracker.Nonce)
+					break
+				}
+				coinType := cctx.InboundParams.CoinType
+
+				txCount := 0
+				var txResult *rpc.GetTransactionResult
+				for _, txHash := range tracker.HashList {
+					if result, ok := ob.CheckFinalizedTx(ctx, txHash.TxHash, nonce, coinType); ok {
+						txCount++
+						txResult = result
+						ob.Logger().Outbound.Info().
+							Msgf("WatchOutbound: confirmed outbound %s for chain %d nonce %d", txHash.TxHash, chainID, nonce)
+						if txCount > 1 {
+							ob.Logger().Outbound.Error().Msgf(
+								"WatchOutbound: checkFinalizedTx passed, txCount %d chain %d nonce %d txResult %v", txCount, chainID, nonce, txResult)
+						}
+					}
+				}
+				// should be only one finalized txHash for each nonce
+				if txCount == 1 {
+					ob.SetTxResult(nonce, txResult)
+				} else if txCount > 1 {
+					// should not happen. We can't tell which txHash is true. It might happen (e.g. glitchy/hacked endpoint)
+					ob.Logger().Outbound.Error().Msgf("WatchOutbound: finalized multiple (%d) outbound for chain %d nonce %d", txCount, chainID, nonce)
+				}
+			}
+			ticker.UpdateInterval(ob.GetChainParams().OutboundTicker, ob.Logger().Outbound)
+		case <-ob.StopChannel():
+			ob.Logger().Outbound.Info().Msgf("WatchOutbound: watcher stopped for chain %d", chainID)
+			return nil
+		}
+	}
+}
+
+// IsOutboundProcessed checks outbound status and returns (isIncluded, isFinalized, error)
+// It also posts vote to zetacore if the tx is finalized
+// TODO(revamp): rename as it also vote the outbound
 func (ob *Observer) IsOutboundProcessed(
```

Comment:
We need to rename this method. This is very confusing to have a IsSomething method doing side effects and not just returning a boolean

Reply:
The interface will be renamed as
`VoteOutboundIfConfirmed(ctx context.Context,	cctx *crosschaintypes.CrossChainTx) (bool, error)`

returning a flag `continueKeysign` to indicate if TSS keysign needs to continue.

Reply:
the refactor is in this commit https://github.com/zeta-chain/node/pull/2560/commits/d40364f3e20902b15f5fb82525de3c9f53c40bbd